File: server/services/media_manager.py
import uuid
from pathlib import Path
from typing import BinaryIO

from infra.log import logger
from infra.s3 import S3Storage

MAX_FILE_SIZE = 500 * 1024 * 1024  # 500 MB

ALLOWED_MIME_PREFIXES = (
    "image/",
    "audio/",
    "video/",
    "application/pdf",
    "application/zip",
    "application/octet-stream",
    "text/",
)


class MediaManager:
    """Manages file upload to S3 object storage."""

    def __init__(self, s3: S3Storage):
        self._s3 = s3

    async def store(
        self,
        file_obj: BinaryIO,
        file_name: str,
        file_size: int,
        mime_type: str,
        session_id: str | None = None,
    ) -> dict | None:
        """Upload a file to S3.

        ``file_obj`` is a seekable file-like object (e.g. ``SpooledTemporaryFile``
        from FastAPI ``UploadFile.file``).  The caller is responsible for
        seeking to the start before calling this method.
        """
        if file_size > MAX_FILE_SIZE:
            logger.warning("Media rejected: file too large ({} bytes, max={})", file_size, MAX_FILE_SIZE)
            return None
        if file_size == 0:
            logger.warning("Media rejected: empty file (name={})", file_name)
            return None
        if not self._is_mime_allowed(mime_type):
            logger.warning("Media rejected: unsupported MIME type '{}' (name={})", mime_type, file_name)
            return None

        # Sanitize filename (prevent path traversal and bare dot-segments)
        safe_name = Path(file_name).name
        if not safe_name or safe_name.startswith("."):
            safe_name = "unnamed"

        # Use provided sessionId or generate a random UUID
        sid = session_id or uuid.uuid4().hex
        # S3 key uses the raw filename; URL encoding is handled by S3Storage
        # when constructing the public download URL.
        key = f"{sid}/{safe_name}"

        download_url = await self._s3.put_object(key, file_obj, mime_type, file_size)

        logger.info(
            "Stored media s3://{}/{} ({}, {} bytes)",
            self._s3.bucket, key, mime_type, file_size,
        )
        return {
            "fileName": safe_name,
            "mimeType": mime_type,
            "fileSize": file_size,
            "sessionId": sid,
            "downloadUrl": download_url,
        }

    def _is_mime_allowed(self, mime_type: str) -> bool:
        if not mime_type:
            return False
        for prefix in ALLOWED_MIME_PREFIXES:
            if mime_type.startswith(prefix):
                return True
        return False

File: backend/internal/service/media_manager.go
package service

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"net/http"
	"path/filepath"
	"strings"

	"github.com/google/uuid"
	"github.com/rs/zerolog/log"

	"astron-claw/backend/internal/infra/storage"
)

const MaxFileSize = 500 * 1024 * 1024 // 500 MB

var (
	ErrFileTooLarge = errors.New("file too large")
	ErrFileEmpty    = errors.New("file is empty")
	ErrMIMERejected = errors.New("unsupported MIME type")
)

var allowedMIMEPrefixes = []string{
	"image/",
	"audio/",
	"video/",
	"application/pdf",
	"application/zip",
	"application/octet-stream",
	"text/",
}

// MediaManager manages file uploads to object storage.
type MediaManager struct {
	storage storage.ObjectStorage
}

// NewMediaManager creates a new MediaManager.
func NewMediaManager(s storage.ObjectStorage) *MediaManager {
	return &MediaManager{storage: s}
}

// MediaResult represents the result of a file upload.
type MediaResult struct {
	FileName    string `json:"fileName"`
	MimeType    string `json:"mimeType"`
	FileSize    int64  `json:"fileSize"`
	SessionID   string `json:"sessionId"`
	DownloadURL string `json:"downloadUrl"`
}

// Store uploads a file to object storage.
func (m *MediaManager) Store(body io.Reader, fileName string, fileSize int64, mimeType, sessionID string) (*MediaResult, error) {
	if fileSize > MaxFileSize {
		log.Warn().Int64("size", fileSize).Int64("max", MaxFileSize).Msg("Media rejected: file too large")
		return nil, ErrFileTooLarge
	}
	if fileSize == 0 {
		log.Warn().Str("name", fileName).Msg("Media rejected: empty file")
		return nil, ErrFileEmpty
	}

	// MIME sniffing: read first 512 bytes to detect actual content type
	buf := make([]byte, 512)
	n, err := io.ReadAtLeast(body, buf, 1)
	if err != nil && err != io.ErrUnexpectedEOF {
		return nil, fmt.Errorf("read for MIME detection: %w", err)
	}
	detected := http.DetectContentType(buf[:n])
	// Use detected type if the declared type is generic
	if mimeType == "" || mimeType == "application/octet-stream" {
		mimeType = detected
	}

	if !isMIMEAllowed(mimeType) {
		log.Warn().Str("mime", mimeType).Str("detected", detected).Str("name", fileName).Msg("Media rejected: unsupported MIME type")
		return nil, ErrMIMERejected
	}

	// Preserve seekability for multipart uploads so S3-compatible SDKs can
	// compute checksums on plain HTTP endpoints such as local MinIO.
	storageBody, err := rewindForStorage(body, buf[:n])
	if err != nil {
		return nil, err
	}

	// Sanitize filename
	safeName := filepath.Base(fileName)
	if safeName == "" || safeName == "." || strings.HasPrefix(safeName, ".") {
		safeName = "unnamed"
	}

	// Use provided sessionId or generate UUID
	sid := sessionID
	if sid == "" {
		sid = uuid.New().String()
	}

	key := sid + "/" + safeName

	downloadURL, err := m.storage.PutObject(key, storageBody, mimeType, fileSize)
	if err != nil {
		return nil, err
	}

	log.Info().
		Str("bucket", m.storage.BucketName()).
		Str("key", key).
		Str("mime", mimeType).
		Int64("size", fileSize).
		Msg("Stored media")

	return &MediaResult{
		FileName:    safeName,
		MimeType:    mimeType,
		FileSize:    fileSize,
		SessionID:   sid,
		DownloadURL: downloadURL,
	}, nil
}

func rewindForStorage(body io.Reader, sniffed []byte) (io.Reader, error) {
	if rs, ok := body.(io.ReadSeeker); ok {
		if _, err := rs.Seek(0, io.SeekStart); err != nil {
			return nil, fmt.Errorf("rewind seekable body: %w", err)
		}
		return rs, nil
	}

	payload, err := io.ReadAll(body)
	if err != nil {
		return nil, fmt.Errorf("buffer non-seekable body: %w", err)
	}
	combined := append(append([]byte{}, sniffed...), payload...)
	return bytes.NewReader(combined), nil
}

func isMIMEAllowed(mimeType string) bool {
	if mimeType == "" {
		return false
	}
	for _, prefix := range allowedMIMEPrefixes {
		if strings.HasPrefix(mimeType, prefix) {
			return true
		}
	}
	return false
}
